### ISSUE TITLE
Preprocessor macros

### ISSUE THREAD

**Issue by EliotVU** [EliotVU/UnrealScript-Language-Service#10]
The usage of macros can lead to severe parsing and indexing issues, like the following code:

```return ( `TimeSince( LastRenderTime ) < 0.3 );```

The usage of `TimeSince here will tell the lexer that this is a PP command, when this is occurred the input will be thrown out thus leaving the parser with:

```return (  < 0.3 );```

Which leads to even bigger issues down the road, so as the end block of a method not being terminated.

One workaround for this issue could be to safely parse binary operators that are missing an expression on the left or right, however due the nature of UnrealScript allowance of custom defined pre and post operators, "< 0.3" could in theory be a post operator! Thus it cannot be assumed to be an incomplete binary operator (albeit it would be totally unexpected for "<" symbol, that cannot be said of other symbols).

The proper solution would be not to skip any PP commands and instead fully parse and evaluate any macro prior to parsing the document, this is however not an easy task!

**Comment by Xymanek** [comment]
Status of marcos in 0.4.3:

Parsing still fails. Sometimes it causes just a small error:
![image](https://user-images.githubusercontent.com/2865341/66703044-e1827f00-ed16-11e9-9113-f5ef25d6d267.png)

Sometimes a much bigger problem:
![image](https://user-images.githubusercontent.com/2865341/66703054-070f8880-ed17-11e9-80d2-0ed71a543fd8.png)
The "problems" window is full of `(ANTLR Node Error) ;`

**Comment by EliotVU** [comment]
Yeah, macros with params are not yet supported, neither are macro symbols being included yet from the global.uci or `include calls.

**Comment by tuokri** [comment]
I was thinking, would it be possible to use the output from `make -intermediate` to improve the plugin's macro parsing?

**Comment by EliotVU** [comment]
@tuokri Yes, or even just displace the workspace folders with macro code with the intermediate code; The only problem with this solution would be for modders that need to mod the stock scripts.

**Comment by EliotVU** [comment]
I have been working on and off on a new approach to tackle this issue in another branch, but nothing complete yet (ANTLR's approach really likes to interfere with most solutions)

**Comment by EliotVU** [comment]
I have worked through the road blocking issues I was having with the re-worked preprocessor, see commit 4aefe32d3e22b30adacf96d40475743304e61316

It can now preprocess:

* Any `include even `include within any `include including all its macro and non-macro tokens!
* Macros with arguments `Log(text...)
* Concatenated macros `{prefix}Identifier

However this is not yet production ready, there's still work to be done in regards to:

* Counter, SetCounter and GetCounter macros.
* Syntax highlighting, proper hoverinfo and goto definition handling.
* Incomplete typed macros
* Macro inheritance and global macros relative to the 'package' directory.
* Cannot parse `{endif}
* Non-macro code between a `if and `endif block are always included when they shouldn't be.

**Comment by DaloLorn** [comment]
Got a possible edge case to look for: Root-level macro expansion.

I've seen a few macros out in the wild that expand out to whole functions, but the current prod parser chokes on the unexpected input. As a result, everything past that line receives only basic syntax highlighting.

A stripped-down example from an XCOM 2 mod I've been looking at:

```
// The language server fully recognizes this function.
simulated function DrawBox (int Chance, string colour, out int offsetX)
{
  // Implementation omitted for brevity
}

`MCM_CH_VersionChecker(class'MCM_Defaults'.default.VERSION, class'WOTC_DisplayHitChance_MCMScreen'.default.CONFIG_VERSION)

/* The above macro expands out to something like this:
function bool MCM_CH_IMPL_VersionChecker()
{
  return (class'MCM_Defaults'.default.VERSION < class'WOTC_DisplayHitChance_MCMScreen'.default.CONFIG_VERSION);
}
function int MCM_CH_IMPL_VersionMax()
{
  return (MCM_CH_IMPL_VersionChecker() ? (class'MCM_Defaults'.default.VERSION) : (class'WOTC_DisplayHitChance_MCMScreen'.default.CONFIG_VERSION));
}
*/

// The language server rejects all invocations of Rename Symbol on this function, which is the first thing I noticed.
// It also reports `'GetCritDamage' does not exist on type 'WOTC_DisplayHitChance.WOTC_DisplayHitChance_UITacticalHUD_ShotHUD.Update'` when the mod tries to use it.
static function int GetCritDamage(XcomGameState_Ability AbilityState, StateObjectReference TargetRef, optional out array<UISummary_ItemStat> ItemsStat) {
  // Implementation omitted for brevity
}
```

**Comment by EliotVU** [comment]
Thanks, that one may possible have been fixed in the 'develop' branch already, but I'll add a unit test for it to be sure.
